### ISSUE TITLE
pyright can't infer type of Unknown variable when told that the value is in a list of strings

### ISSUE THREAD

**Issue by nikita-kiryanov** [microsoft/pyright#2169]
I have code that obtains an `error_value` from a third party function. According to pyright, the value of this `error_value` is **Unknown**. The following code fails to infer the actual type of `error_value`, even though the if statement makes it clear that it should be a string.
```python
def foo(self, error_value):  # error_value type is Unknown
        error_codes: Dict[str, int] = {'a': 1, 'b': 2, 'c': 3}
        if error_value in list(error_codes.keys()):
            self.error.set_int_error_code(error_codes[error_value])  # Type checker error here
```

The error I get is this:

> Argument of type "Unknown | None" cannot be assigned to parameter "k" of type "str" in function "\_\_getitem\_\_"
>   Type "Unknown | None" cannot be assigned to type "str"
>     Type "None" cannot be assigned to type "str"
> (variable) error_value: Unknown | None

**Expected behavior**
pyright should understand that if a value is `in` a list of strings, then it's a string.

**VS Code extension or command-line**
I'm using pyright via the Pylance extension, which is version v2021.8.0.

**Comment by nikita-kiryanov** [comment]
I'm not on strict mode.

Anyway, as I was constructing a standalone example I figured out that what was bothering the type checker wasn't Unknown, but rather the fact that the return value was "Unknown | **None**". It's the `None` part it had a problem with.

The function that supplied the problematic value was this:
```python
def get_client_data(self, idx):
        if idx == -1:
            return None

        return self.this_function_returns_unknown(idx)
```

If I modify the if statement in the original example as follows:
```python
if error_value in list(error_codes.keys()) and error_value is not None:
```
...then pyright doesn't complain anymore.

**Comment by erictraut** [comment]
OK, sounds like it's working as expected then. I'm going to close the issue. If you have any other questions, let us know.
